feat(misskon): add source misskon.

marioplus opened the pull request:
refs: #1855

Checklist:

- [x] Updated `extVersionCode` value in `build.gradle` for individual extensions
- [x] Updated `overrideVersionCode` or `baseVersionCode` as needed for all multisrc extensions
- [x] Referenced all related issues in the PR body (e.g. "Closes #xyz")
- [x] Added the `isNsfw = true` flag in `build.gradle` when appropriate
- [ ] Have not changed source names
- [ ] Have explicitly kept the `id` if a source's name or language were changed
- [x] Have tested the modifications by compiling and running the extension through Android Studio
- [x] Have removed `web_hi_res_512.png` when adding a new extension

marioplus committed:
feat(misskon): add source misskon.

refs: #1855

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+        manga.title = titleEL.text()
+        manga.thumbnail_url = element.select(".post-thumbnail img").attr("data-src")
+        manga.setUrlWithoutDomain(titleEL.select("a").attr("href").substring(baseUrl.length))
+        return manga
```
Use `SManga.create().apply { ... }`

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+
+        val manga = SManga.create()
+        manga.title = titleEL.text()
+        manga.thumbnail_url = element.select(".post-thumbnail img").attr("data-src")
```
Use `.selectFirst` and `.absUrl`

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+        val manga = SManga.create()
+        manga.title = titleEL.text()
+        manga.thumbnail_url = element.select(".post-thumbnail img").attr("data-src")
+        manga.setUrlWithoutDomain(titleEL.select("a").attr("href").substring(baseUrl.length))
```
Don't need `.substring`, `setUrlWithoutDomain` does this for you. You should in addition use `.absUrl` so the responsibility of catering for relative URLs, weird schema syntax etc. is handled by Jsoup rather than `setUrlWithoutDomain` -- which does not have any additional context.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+    override fun simpleNextPageSelector(): String? = null
+
+    // region popular
+    override fun popularMangaRequest(page: Int) = GET("$baseUrl/top3/")
```
```suggestion
    override fun popularMangaRequest(page: Int) = GET("$baseUrl/top3/", headers)
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+            "$baseUrl/page/$page"
+        }
+        return GET(url)
+    }
```
```kt
return GET(baseUrl + if (page != 1) "/page/$page" else "", headers)
```

`page` is 1-indexed. There are many variants of this check across the repository.
```
page > 1
page == 1
page != 1
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+        return if (filter.isSelected()) {
+            GET(filter.selected.url)
+        } else {
+            GET("$baseUrl/page/$page/?s=$query")
```
Use `HttpUrl` to build a query, so the query is properly escaped. Also add `headers`.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+
+    override fun chapterFromElement(element: Element): SChapter {
+        val dataSrc = element.selectFirst(".entry img")!!.attr("data-src")
+        val dataRegex = Regex("^.+(\\d{4}/\\d{2}/\\d{2}).+$")
```
Extract regex out to class or companion object.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+        val manga = SManga.create()
+        manga.title = postInnerEl.select(".post-title").text()
+        manga.description = ""
+        manga.genre = postInnerEl.select(".post-tag > a").joinToString(", ") { it.text() }
```
```suggestion
        manga.genre = postInnerEl.select(".post-tag > a").joinToString { it.text() }
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+
+        val manga = SManga.create()
+        manga.title = postInnerEl.select(".post-title").text()
+        manga.description = ""
```
```suggestion
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+    override fun mangaDetailsParse(document: Document): SManga {
+        val postInnerEl = document.select("article > .post-inner")
+
+        val manga = SManga.create()
```
Also use `.apply {}` here.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+
+    // region Details
+    override fun mangaDetailsParse(document: Document): SManga {
+        val postInnerEl = document.select("article > .post-inner")
```
I assume you only expect one object here, should thus use `.selectFirst`

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+    override fun chapterListSelector() = "html"
+
+    override fun chapterFromElement(element: Element): SChapter {
+        val dataSrc = element.selectFirst(".entry img")!!.attr("data-src")
```
This is only used for the date, which is an optional field. Should not use non-null assert.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+            "1997/01/01"
+        }
+
+        val chapter = SChapter.create()
```
Also use `.apply {}` here.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+
+        val chapter = SChapter.create()
+        chapter.chapter_number = 0F
+        chapter.setUrlWithoutDomain(element.selectFirst("link[rel=\"canonical\"]")!!.attr("href"))
```
```suggestion
        chapter.setUrlWithoutDomain(element.selectFirst("link[rel=\"canonical\"]")!!.absUrl("href"))
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+        chapter.chapter_number = 0F
+        chapter.setUrlWithoutDomain(element.selectFirst("link[rel=\"canonical\"]")!!.attr("href"))
+        chapter.name = dateStr
+        chapter.date_upload = SimpleDateFormat("yyyy/MM/dd", Locale.ENGLISH).parse(dateStr)!!.time
```
Extract to class or companion object.

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+                }
+                doc.select("div.post-inner > div.entry > p > img")
+                    .map { it.attr("data-src") }
+                    .forEach { pages.add(Page(pages.size, "", it)) }
```
```suggestion
                    .forEach { pages.add(Page(pages.size, imageUrl = it)) }
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+                    }
+                }
+                doc.select("div.post-inner > div.entry > p > img")
+                    .map { it.attr("data-src") }
```
```suggestion
                    .map { it.absUrl("data-src") }
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,258 @@
+
+    // region Pages
+    override fun pageListParse(document: Document): List<Page> {
+        val basePageUrl = document.selectFirst("link[rel=\"canonical\"]")!!.attr("href")
```
```suggestion
        val basePageUrl = document.selectFirst("link[rel=\"canonical\"]")!!.absUrl("href")
```

marioplus committed:
feat(misskon): Improve network handling and code structure

• Network enhancements:
  - Add headers parameter for `GET` methods
  - Implement standardized query with `HttpUrl`
  - Replace `attr()` with `asbUrl()` for safe path composition

• Structural improvements:
  - Move date regex to companion object
  - Extract SimpleDateFormat as global constant
  - Refactor Filter logic into standalone class

• Robustness upgrades:
  - Remove unsafe non-null assertion(!!) on date_upload
  - Adopt SManga.create().apply  { } chaining pattern

refs: #1855

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,135 @@
+    override fun simpleMangaFromElement(element: Element) = SManga.create().apply {
+        val titleEL = element.select(".post-box-title")
+        title = titleEL.text()
+        thumbnail_url = element.selectFirst(".post-thumbnail img")!!.absUrl("data-src")
```
`thumbnail_url` is not critical, should remove non-null assert.
```suggestion
        thumbnail_url = element.selectFirst(".post-thumbnail img")?.absUrl("data-src")
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,135 @@
+
+    override fun simpleMangaSelector() = "article.item-list"
+    override fun simpleMangaFromElement(element: Element) = SManga.create().apply {
+        val titleEL = element.select(".post-box-title")
```
You only expect one element here.
```suggestion
        val titleEL = element.selectFirst(".post-box-title")
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,135 @@
+        val titleEL = element.select(".post-box-title")
+        title = titleEL.text()
+        thumbnail_url = element.selectFirst(".post-thumbnail img")!!.absUrl("data-src")
+        setUrlWithoutDomain(titleEL.select("a").attr("abs:href"))
```
You only expect one element here.
```suggestion
        setUrlWithoutDomain(titleEL.selectFirst("a").attr("abs:href"))
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,135 @@
+            GET(it.url, headers)
+        } ?: run {
+            baseUrl.toHttpUrl().newBuilder()
+                .addPathSegment("/page/$page/")
```
```suggestion
            "$baseUrl/page/$page/".toHttpUrl().newBuilder()
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,135 @@
+    // region Details
+    override fun mangaDetailsParse(document: Document) = SManga.create().apply {
+        val postInnerEl = document.select("article > .post-inner")
+        title = postInnerEl.select(".post-title").text()
```
You only expect one element here.
```suggestion
        title = postInnerEl.selectFirst(".post-title").text()
```

vetleledaal commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,135 @@
+                    }
+                }
+                doc.select("div.post-inner > div.entry > p > img")
+                    .map { it.attr("data-src") }
```
Did using `.absUrl` here break it?

marioplus committed:
feat(misskon): improve code structure and null safety

- Replace `attr()` with `absUrl()` for more reliable URL extraction
- Remove non-null assertion (`!!`) from `thumbnail_url` as it's not critical
- Fix usage of `select()` and `selectFirst()` to properly handle nullable cases

refs: #1855

marioplus committed:
Merge branch 'main' into misskon

marioplus committed:
chore(MissKon): remove needless blank line

- Fix lint violation (no-consecutive-blank-lines)

Closes: #1855

marioplus committed:
Merge branch 'misskon' of https://github.com/marioplus/extensions-source into misskon

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,139 @@
+            chapter_number = 0F
+            setUrlWithoutDomain(element.selectFirst("""link[rel="canonical"]""")!!.absUrl("href"))
+            name = dateStr
+            date_upload = FULL_DATE_FORMAT.parse(dateStr)!!.time
```
```suggestion
            date_upload = FULL_DATE_FORMAT.tryParse(dateStr)
```
need to import too

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,139 @@
+        SourceCategorySelector.create(),
+    )
+
+    private inline fun <reified T> Iterable<*>.findInstance() = find { it is T } as? T
```
this is available in `keiyoushi.utils.firstInstance`

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,139 @@
+            ?.let { url ->
+                FULL_DATE_REGEX.find(url)?.groupValues?.get(1)
+                    ?: YEAR_MONTH_REGEX.find(url)?.groupValues?.get(1)?.let { "$it/01" }
+            } ?: DEFAULT_DATE_STR
```
dont fall back to unix time start, just dont set a date

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,139 @@
+        return SChapter.create().apply {
+            chapter_number = 0F
+            setUrlWithoutDomain(element.selectFirst("""link[rel="canonical"]""")!!.absUrl("href"))
+            name = dateStr
```
just set "Gallery" or similar as name

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/SourceCategorySelector.kt:
```diff
@@ -0,0 +1,128 @@
+
+            val TOP = listOf(
+                SourceCategory("Top 3 days", "/top3/"),
+                SourceCategory("Top 7 days", "https://misskon.com/top7/"),
```
imo dont set baseurl in here, and build the full url in `searchMangaRequest`

cuong-tran commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,139 @@
+    // endregion
+
+    // region latest
+    override fun latestUpdatesRequest(page: Int) = GET(baseUrl + if (page != 1) "/page/$page" else "", headers)
```
although `$baseUrl/page/1` get redirect to `$baseUrl/`, it should still work

marioplus committed:
refactor(MissKon): optimize URL handling and reuse utils

* Remove baseUrl from category URLs (build dynamically when used)
* Reuse existing utils for:
  - Date parsing
  - Filter queries
* Standardize pagination URL construction

Closes: #1855

cuong-tran commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -48,16 +50,16 @@ class MissKon() : SimpleParsedHttpSource() {
 
     // region latest
-    override fun latestUpdatesRequest(page: Int) = GET(baseUrl + if (page != 1) "/page/$page" else "", headers)
+    override fun latestUpdatesRequest(page: Int) = GET("#baseUrl/page/$page", headers)
```
```suggestion
    override fun latestUpdatesRequest(page: Int) = GET("$baseUrl/page/$page", headers)
```

cuong-tran commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -86,20 +88,20 @@ class MissKon() : SimpleParsedHttpSource() {
-            name = dateStr
-            date_upload = FULL_DATE_FORMAT.parse(dateStr)!!.time
+            setUrlWithoutDomain(element.selectFirst("link[rel=canonical]")!!.absUrl("href"))
+            name = dateStr ?: "Gallery"
```
```suggestion
            name = "Gallery"
```

cuong-tran commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,137 @@
+    }
+
+    override fun searchMangaNextPageSelector() = "div.content > div.pagination > span.current + a"
+    override fun searchMangaSelector() = "article.item-list"
```
same as `simpleMangaSelector`

marioplus committed:
fix(MissKon): Correct URL template, set default gallery name, and unify selectors

Closes: #1855

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,136 @@
+    override fun searchMangaRequest(page: Int, query: String, filters: FilterList): Request {
+        val filter = filters.firstInstance<SourceCategorySelector>()
+        return filter.selectedCategory?.let {
+            GET("$baseUrl$it.url", headers)
```
```suggestion
            GET("$baseUrl${it.url}", headers)
```
you are putting literal string "url" in the url?

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,136 @@
+
+    override val client = network.cloudflareClient.newBuilder()
+        .rateLimitHost(baseUrl.toHttpUrl(), 10, 1, TimeUnit.SECONDS)
+        .setRandomUserAgent(UserAgentType.MOBILE)
```
is this necessary, from a quick look on desktop, the site works/have similar html structure as mobile site, so shouldn't limit it to one specific useragent type

AwkwardPeak7 commented on src/all/misskon/src/eu/kanade/tachiyomi/extension/all/misskon/MissKon.kt:
```diff
@@ -0,0 +1,136 @@
+        val postInnerEl = document.selectFirst("article > .post-inner")!!
+        return SManga.create().apply {
+            title = postInnerEl.select(".post-title").text()
+            genre = postInnerEl.select(".post-tag > a").joinToString { it.text() }
```
also set: `update_strategy = UpdateStrategy.ONLY_FETCH_ONCE`

marioplus committed:
refactor(MissKon): fix URL construction and client configuration

- Fix string interpolation in search URL ("$it.url"  → "${it.url}")
- Remove MOBILE user agent restriction
- Remove unused dependency
- Add ONLY_FETCH_ONCE update strategy
- Clean up selector syntax
